### ISSUE TITLE
Introduce multiple return type

### ISSUE THREAD

**Issue by Smit-create** [lcompilers/lpython#379]
We should support `lpython` to allow multiple return types. One example of this is the `pow` function. We know that `pow(x: i32, y: i32)` will return `i32` if `y>=0` else it should return `f64`. So, I am thinking about how should we design the return type notation in this case:

Some of the designs are:

```py
def pow(x: i32, y: i32) -> Union[i32, f32]:
    pass
```
or

```py
def pow(x: i32, y: i32) -> i32 | f32:
    pass
```

Let me know you views and suggestions.

**Comment by Madhav2310** [comment]
I like the latter more, looks elegant. Also, it's more predominantly used so will be easier to understand by developers.

**Comment by certik** [comment]
This is more tricky than it might seem. To allow `Union` in a performant manner, it would be a generic (template) feature. The code would get instantiated once it is called, and specialized for the given types. This can be done for things like `def f(x: i32 | f32)`. However, as a return type that gets tough --- one could infer that from where it is assigned to, but that might not always be possible if you for example have another generic function `def g(y: i32 | f32)` and you do `g(pow(..))`, then you do not know the types, and thus what function to call.

The other way to implement `Union` is via runtime generics, for example you could represent the `i32 | f32` as a struct (actual C style union) with a tag to know the type. That will work, but would not be performing well. I think we should not go that route. 

The good news is that I think the builtin function `pow` is rarely used. So we could simply always return `f32`. And one can do something like `x: i32 = int(pow(2, 3))`, which will work in both CPython and LPython and behave the same way. And be performing.

More important is what to do about regular `**` operator which behaves the same way:

```
>>> 3**2
9
>>> 3**(-2)
0.1111111111111111
```

Well, if we know the values at compile time, we can figure out the type. But what about this:
```
>>> x = 3
>>> y = 2
>>> x**y
9
>>> y = -2
>>> x**y
0.1111111111111111
```
The type truly depends on the runtime values of `y`, which could be read from an input file, thus we do not know at compile time what the type will be.

Here is one way out:

* For integer x,y, we always return an integer for positive `y`, and for negative `y` we return a runtime exception in Debug mode. That way you would know that the feature is not supported. But in the code you can assume the result is an integer. You test your code and ensure you get no runtime exceptions in Debug mode. Then you recompile in Release mode and you will not have any checks and thus things will be fast. The contract of LPython is that if it compiles and you get no runtime errors in Debug mode, the behavior will be equivalent as CPython.

* For floating point x,y, you always return a floating point number. That is the same behavior as in CPython, so no problem.

**Comment by czgdp1807** [comment]
I am confused here. Why to return anything when both of the inputs are integers and `y < 0`. LPython is statically typed version of Python. I think LPython should prefer numpy behaviour more than the Python behaviour. From this I mean if `pow(x: i32, y: i32)` should raise an error when `y < 0` and if someone wants to pass `y < 0` then just use floats as inputs. See the corresponding numpy example below,

```python
>>> import numpy as np
>>> np.power(int(1), -int(2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Integers to negative integer powers are not allowed.
>>> np.power(float(1), -float(2))
1.0
```

P.S. Its totally my personal opinion.

**Comment by czgdp1807** [comment]
> one could infer that from where it is assigned to, but that might not always be possible if you for example have another generic function `def g(y: i32 | f32)` and you do `g(pow(..))`, then you do not know the types, and thus what function to call.

I think the number of combinations of different types possible will increase exponentially. So you will have to generate code for each of those combinations possible. A big scaling problem I feel.

> For integer x,y, we always return an integer for positive y, and for negative y we return a runtime exception in Debug mode.

I strongly think that's the way to go. Simple and effective. No need to overthink about what types will be returned. Got an integer, return an integer, can't return it? just raise an error. :-).

**Comment by certik** [comment]
Yes. We should start just by supporting a subset (but a very usable subset) and give a compile time (or runtime) error for everything else. I think being the most general here is not the bottleneck that we have to focus right now. Rather, we should focus on NumPy's operations and implementation.
